airflow/providers/openlineage/utils/__init__.py

```diff
@@ -0,0 +1,499 @@
+            )
+        )
+    }
+
+
+def get_dagrun_start_end(dagrun: DagRun, dag: DAG):
+    try:
+        return dagrun.data_interval_start, dagrun.data_interval_end
+    except AttributeError:
+        # Airflow < 2.2 before adding data interval
+        pass
+    start = dagrun.execution_date
+    end = dag.following_schedule(start)
+    return start, end or start
+
+
+def import_from_string(path: str):
+    try:
+        module_path, target = path.rsplit(".", 1)
+        module = importlib.import_module(module_path)
+        return getattr(module, target)
+    except Exception as e:
+        log.warning(e)
+        raise ImportError(f"Failed to import {path}") from e
+
+
+def try_import_from_string(path: str):
+    try:
+        return import_from_string(path)
+    except ImportError:
+        return None
+
+
+def redact_with_exclusions(source: Any):
+    try:
+        from airflow.utils.log.secrets_masker import (
+            _secrets_masker,
+            should_hide_value_for_key,
+        )
+    except ImportError:
+        return source
+    import copy
+
+    sm = copy.deepcopy(_secrets_masker())
+    MAX_RECURSION_DEPTH = 20
+
+    def _redact(item, name: str | None, depth: int):
+        if depth > MAX_RECURSION_DEPTH:
+            return item
+        try:
+            if name and should_hide_value_for_key(name):
+                return sm._redact_all(item, depth)
+            if isinstance(item, dict):
+                return {
+                    dict_key: _redact(subval, name=dict_key, depth=(depth + 1))
+                    for dict_key, subval in item.items()
+                }
+            elif is_dataclass(item) or (is_json_serializable(item) and hasattr(item, "__dict__")):
+                for dict_key, subval in item.__dict__.items():
+                    if _is_name_redactable(dict_key, item):
+                        setattr(
+                            item,
+                            dict_key,
+                            _redact(subval, name=dict_key, depth=(depth + 1)),
+                        )
+                return item
+            elif isinstance(item, str):
+                if sm.replacer:
+                    return sm.replacer.sub("***", item)
+                return item
+            elif isinstance(item, (tuple, set)):
+                return tuple(_redact(subval, name=None, depth=(depth + 1)) for subval in item)
+            elif isinstance(item, list):
+                return [_redact(subval, name=None, depth=(depth + 1)) for subval in item]
+            else:
+                return item
+        except Exception as e:
+            log.warning(
+                "Unable to redact %s" "Error was: %s: %s",
+                repr(item),
+                type(e).__name__,
+                str(e),
+            )
+            return item
+
+    return _redact(source, name=None, depth=0)
+
+
+def is_dataclass(item):
+    return getattr(item.__class__, "__attrs_attrs__", None) is not None
```

Comment:
a) this is not dataclass, but attrs, b) https://www.attrs.org/en/stable/api.html#attrs.has

```suggestion
    return attrs.has(item.__class__)   
```

Reply:
Done